### DISCUSSION TITLE
Google Cloud Spanner support

### DISCUSSION THREAD

**Discussion by amacneil** [amacneil/dbmate#369] [Feature Ideas]
Need to add support for Google Cloud Spanner

**Comment by AJCStriker** [comment]
Hey @amacneil I would be interested in picking this issue up and submitting a PR.

I took a look at the current drivers but is there anything to keep in mind whilst designing an integration? 
In particular clarification about how you view Connection lifecycles would be really helpful?

**Comment by amacneil** [comment]
Last time I looked into this, I ran into the issue that dbmate has been designed around an assumption that all drivers will use `database/sql`, and the spanner driver did not use that. So it would require refactoring the driver interface to remove dependence on `database/sql`.

Long term I think this is a good direction to take dbmate, but just be aware that to achieve this feature I think you will need to refactor the interface and all other drivers to match the new interface, it won't be a simple case of adding one new driver.

From memory there also might be a problem with the spanner driver not accepting multiple sql statements in one string, so we will need to import a sql parsing library and do our own statement splitting. Again I think this is a nice feature that would benefit all drivers eventually (allowing us to give more helpful error messages, etc), but it adds complexity.

**Reply by mii9000** [reply]
any plans for this development? i am open to contribute if you can share the design on how it should be done

**Reply by dossy** [reply]
No plans, but it should be much easier now, as this suggests that there's a PostgreSQL client interface to Google Spanner:

https://cloud.google.com/spanner/docs/postgresql-interface

> Developers can connect their applications to a PostgreSQL interface database using open source Spanner clients for Java, Go, Python, Node.js, Ruby, PHP, C#, and C++ environments. They can also use the open source Spanner [JDBC driver](https://github.com/googleapis/java-spanner-jdbc) and the [driver for Go's database/sql package](https://github.com/googleapis/go-sql-spanner).

And, there's also a [Google Cloud Spanner](https://cloud.google.com/spanner) driver for Go's [database/sql](https://golang.org/pkg/database/sql/) package.

https://github.com/googleapis/go-sql-spanner

I don't use and have no plans to use Google Cloud Spanner databases, but anyone who does who wants to work on support for it could discuss it here and see how much interest there is, I suppose.

**Reply by grahamhoyes** [reply]
Here's a PR with a first attempt at adding support for the postgres interface, following the details of my post below: https://github.com/amacneil/dbmate/pull/555

**Comment by grahamhoyes** [comment]
Spanner has two supported dialects at the moment: [GoogleSQL](https://cloud.google.com/spanner/docs/reference/standard-sql/overview) and [Postgres](https://cloud.google.com/spanner/docs/postgresql-interface), the latter with some [limitations](https://cloud.google.com/spanner/docs/reference/postgresql/known-issues-postgresql-interface) and a smaller subset of [supported functions](https://cloud.google.com/spanner/docs/reference/postgresql/functions-and-operators). The dialect for a database is chosen at database creation, and can't be changed - so full-featured support would require separate support for both.

Supporting GoogleSQL would probably mean a new driver that uses https://github.com/googleapis/go-sql-spanner, and the queries would be different from the Postgres driver.

Supporting the Postgres interface through the [PGAdapter](https://cloud.google.com/spanner/docs/pgadapter) should be relatively simple:
- The `unnest(text[])` function used [here](https://github.com/amacneil/dbmate/blob/9b95d4c2c6fc30b641320f2e11ab8774d1052444/pkg/driver/postgres/postgres.go#L451) isn't supported, so like in [this PR that added redshift support](https://github.com/amacneil/dbmate/pull/488/files#diff-9b130c942846428321d903a43dfe2a3ca9598c90a80229490ced07b9d77c4a23R440) we can do identifier quoting client-side
- Spanner doesn't allow migrations in explicit transactions. Setting `transaction:false` in the migrate file works, or we may want to default this to true (while still allowing setting `transaction:true` for data migrations and such)
